### PR TITLE
test: Optimize feature_block_reward_reallocation.py

### PR DESCRIPTION
**Pull request by PastaPastaPasta**
Locally, develop is running in 308 seconds while this branch is running at 199 seconds. A nice improvement if I do say so myself :D

### CODE REVIEW

**Comment by UdjinM6** [comment]
Can confirm 340s -> 130s improvement locally here 👍 I remember that we had to do `sync_blocks` every so often to make sure nodes aren't overwhelmed with blocks which was critical for Travis builds at some point, waiting for https://travis-ci.com/github/UdjinM6/dash/builds/186965759 to confirm that I was too cautious :) Also, pls see 4abd311c98 and 902c35156d

**Comment by PastaPastaPasta** [comment]
I don't see the point of 902c35156d96b081f6f35c54cdb7b436c1cb93d4

**Comment by UdjinM6** [comment]
We push all blocks and check chain state changes via node0 only. Would probably make sense to make sure blocks are propagated via p2p and accepted on the other node too, just in case.

**Comment by PastaPastaPasta** [comment]
Maybe I'm misunderstanding something, but in my mind, this other node is really just needed to serve as the mn to get paid in block reward, there should be no situation where we accept a block (in test suite) where our peer doesn't since we are running the same build and generating / submitblock will validate the block against validation rules in the same way the peer would..

Maybe I'm missing some value

**Comment by UdjinM6** [comment]
>generating / submitblock will validate the block against validation rules in the same way the peer would

But, technically, these are two different routes (rpc vs p2p). If we would need to test internal validation rules only we could probably be simply using unit tests instead.

File: test/functional/feature_block_reward_reallocation.py [hunk 1]
```diff
@@ -3,7 +3,9 @@
 # Distributed under the MIT software license, see the accompanying
 # file COPYING or http://www.opensource.org/licenses/mit-license.php.
 from test_framework.blocktools import create_block, create_coinbase, get_masternode_payment
-from test_framework.mininode import *
+from test_framework.mininode import P2PDataStore, network_thread_start
+from test_framework.messages import CTxOut, FromHex, CCbTx, CTransaction, ToHex
+
```
Comment (PastaPastaPasta) [inline test/functional/feature_block_reward_reallocation.py:8]:
```suggestion
```

**Comment by PastaPastaPasta** [comment]
Seems fair, cherry-picked

**Review by UdjinM6 — APPROVED** [review]
ACK

File: test/functional/feature_block_reward_reallocation.py [hunk 2]
```diff
@@ -15,9 +16,10 @@
 
 '''
 
+
```
Comment (xdustinface) [inline test/functional/feature_block_reward_reallocation.py:19]:
```suggestion
```

File: test/functional/feature_block_reward_reallocation.py [hunk 3]
```diff
@@ -109,38 +106,36 @@ def run_test(self):
         assert_equal(bi['bip9_softforks']['realloc']['status'], 'started')
         assert_equal(bi['bip9_softforks']['realloc']['statistics']['threshold'], self.threshold(0))
 
-        self.signal(399, False) # 1 block short
+        self.signal(399, False)  # 1 block short
```
Comment (xdustinface) [inline test/functional/feature_block_reward_reallocation.py:109]:
```suggestion
        self.signal(399, False) # 1 block short
```

File: test/functional/feature_block_reward_reallocation.py [hunk 4]
```diff
@@ -109,38 +106,36 @@ def run_test(self):
...
 
         self.log.info("Still STARTED but new threshold should be lower at height = 999")
         bi = self.nodes[0].getblockchaininfo()
         assert_equal(bi['blocks'], 999)
         assert_equal(bi['bip9_softforks']['realloc']['statistics']['threshold'], self.threshold(1))
 
-        self.signal(398, False) # 1 block short again
+        self.signal(398, False)  # 1 block short again
```
Comment (xdustinface) [inline test/functional/feature_block_reward_reallocation.py:116]:
```suggestion
        self.signal(398, False) # 1 block short again
```

File: test/functional/feature_block_reward_reallocation.py [hunk 5]
```diff
@@ -109,38 +106,36 @@ def run_test(self):
...
 
         self.log.info("Still STARTED but new threshold should be even lower at height = 1499")
         bi = self.nodes[0].getblockchaininfo()
         assert_equal(bi['blocks'], 1499)
         assert_equal(bi['bip9_softforks']['realloc']['statistics']['threshold'], self.threshold(2))
         pre_locked_in_blockhash = bi['bestblockhash']
 
-        self.signal(396, True) # just enough to lock in
+        self.signal(396, True)  # just enough to lock in
```
Comment (xdustinface) [inline test/functional/feature_block_reward_reallocation.py:124]:
```suggestion
        self.signal(396, True) # just enough to lock in
```

File: test/functional/feature_block_reward_reallocation.py [hunk 6]
```diff
@@ -109,38 +106,36 @@ def run_test(self):
...
-        self.sync_blocks()
 
         self.log.info("Still LOCKED_IN at height = 2498")
         bi = self.nodes[0].getblockchaininfo()
         assert_equal(bi['blocks'], 2498)
         assert_equal(bi['bip9_softforks']['realloc']['status'], 'locked_in')
 
         self.log.info("Advance from LOCKED_IN to ACTIVE at height = 2499")
-        self.nodes[0].generate(1) # activation
+        self.nodes[0].generate(1)  # activation
```
Comment (xdustinface) [inline test/functional/feature_block_reward_reallocation.py:138]:
```suggestion
        self.nodes[0].generate(1) # activation
```

File: test/functional/feature_block_reward_reallocation.py [hunk 7]
```diff
@@ -152,18 +147,16 @@ def run_test(self):
...
         self.nodes[0].generate(9)
-        self.sync_blocks()
         bt = self.nodes[0].getblocktemplate()
         assert_equal(bt['masternode'][0]['amount'], get_masternode_payment(bt['height'], bt['coinbasevalue'], 2500))
         assert_equal(bt['coinbasevalue'], 13748571607)
         assert_equal(bt['masternode'][0]['amount'], 6874285801) # 0.4999999998
 
         self.log.info("Reallocation should kick-in with the superblock mined at height = 2010")
-        for period in range(19): # there will be 19 adjustments, 3 superblocks long each
+        for period in range(19):  # there will be 19 adjustments, 3 superblocks long each
```
Comment (xdustinface) [inline test/functional/feature_block_reward_reallocation.py:156]:
```suggestion
        for period in range(19): # there will be 19 adjustments, 3 superblocks long each
```

File: test/functional/feature_block_reward_reallocation.py [hunk 8]
```diff
@@ -172,36 +165,38 @@ def run_test(self):
         assert_equal(bt['masternode'][0]['amount'], 6132959502) # 0.6
 
         self.log.info("Reward split should stay ~60/40 after reallocation is done")
-        for period in range(10): # check 10 next superblocks
+        for period in range(10):  # check 10 next superblocks
```
Comment (xdustinface) [inline test/functional/feature_block_reward_reallocation.py:168]:
```suggestion
        for period in range(10): # check 10 next superblocks
```

File: test/functional/feature_block_reward_reallocation.py [hunk 9]
```diff
@@ -172,36 +165,38 @@ def run_test(self):
...
+        self.nodes[0].submitblock(ToHex(test_block))
         bi = self.nodes[0].getblockchaininfo()
         assert_equal(bi['blocks'], 1499)
         assert_equal(bi['bip9_softforks']['realloc']['status'], 'started')
         assert_equal(bi['bip9_softforks']['realloc']['statistics']['threshold'], self.threshold(2))
 
         self.log.info("Check thresholds reach min level and stay there")
-        for i in range(8): # 7 to reach min level and 1 more to check it doesn't go lower than that
-            self.signal(0, False) # no need to signal
+        for i in range(8):  # 7 to reach min level and 1 more to check it doesn't go lower than that
```
Comment (xdustinface) [inline test/functional/feature_block_reward_reallocation.py:192]:
```suggestion
        for i in range(8): # 7 to reach min level and 1 more to check it doesn't go lower than that
```

File: test/functional/feature_block_reward_reallocation.py [hunk 10]
```diff
@@ -172,36 +165,38 @@ def run_test(self):
...
+            self.signal(0, False)  # no need to signal
```
Comment (xdustinface) [inline test/functional/feature_block_reward_reallocation.py:193]:
```suggestion
            self.signal(0, False) # no need to signal
```

File: test/functional/feature_block_reward_reallocation.py [hunk 11]
```diff
@@ -172,36 +165,38 @@ def run_test(self):
...
             bi = self.nodes[0].getblockchaininfo()
             assert_equal(bi['blocks'], 1999 + i * 500)
             assert_equal(bi['bip9_softforks']['realloc']['status'], 'started')
             assert_equal(bi['bip9_softforks']['realloc']['statistics']['threshold'], self.threshold(i + 3))
         assert_equal(bi['bip9_softforks']['realloc']['statistics']['threshold'], 300)
 
+
```
Comment (xdustinface) [inline test/functional/feature_block_reward_reallocation.py:200]:
```suggestion
```

**Review by xdustinface — CHANGES_REQUESTED** [review]
I guess you didn't add those intentionally?

**Comment by PastaPastaPasta** [comment]
Those changes are intentional in order to make the file more in line with python formatting standards

**Review by xdustinface — APPROVED** [review]
Ok well then its fine.. it just looked like a mistake lol.. i guess you should (at least try to) split such changes out in a separate commits to make it obvious or write it down somewhere in the description. 

utACK

**Pull request merged**